### PR TITLE
[FLINK-32469][REST] Improve checkpoint REST APIs for programmatic access

### PR DESCRIPTION
**Pull request by hlteoh37**
## What is the purpose of the change

Simplify the checkpoint handler API.
The checkpoint endpoints do not require the entire ExecutionGraph to produce their result. In fact, they only require the CheckpointStatsSnapshot. This change refactors the handlers to use only the CheckpointStatsSnapshot.

## Brief change log
- Checkpoint REST handlers now consume from `AbstractCheckpointStatsHandler` that provides the `CheckpointStatsSnapshot`.
- `CheckpointsStatsSnapshot` is cached, to reduce call rate to underlying JobMaster. 

## Verifying this change
This change added tests and can be verified as follows:
  - Added unit tests
  - Manually ran the Flink dashboard and tested the refresh rate on checkpoint API.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by hlteoh37** [comment]
## Performing performance test on `/checkpoints` API

From the below tests, we ran the following:
1. Control test where we don't have changes to Flink REST API. ExecutionGraph cache of 30s
2. Checkpoint cache with 3s cache
3. No checkpoint cache

We can see that there is no significant differences in the access pattern / latency pattern of the REST API for `/checkpoints`.

We also found the following nuances:
- Periodic `500` returned by `/checkpoints` on the first call after a job restart (likely that there are no "Quantile" statistics after only 1 checkpoint. **This is consistent for all 3 tests, meaning its an independent issue.**
- p95 access pattern seems spiky. Seems like averages around 15ms, but goes up to 40 / 50ms occasionally. **This pattern is consistent across all tests, so this is not a problem introduced by latest change.**


### Test configuration
- Cluster setup
  - Single JM with Zookeeper HA
  - Running on K8s
  - JM resources (vCPU: 200m, Memory: 6083Mi)
  - ResourceManager, JobMaster, BlobServer all running on the same JM container.
- Test setup
  - TPS: 10 req/s
  - Endpoint `/jobs/:jobid/checkpoints`
  - Simple Flink job reading from KDS writing to KDS
  - Checkpoint interval 1s, min pause between checkpoints 1s

### Control Test 1 (no changes to Flink REST API): ExecutionGraph cache of 3s, Job restarting constantly

Screenshot 2023-07-03 at 11 21 01


JM CPU/memory use

```
NAME                                         CPU(cores)   MEMORY(bytes)
flink-jobmanager-dc5d8b5df-875x6             51m          1271Mi
```
Screenshot 2023-07-03 at 11 20 54


### Test 2: 3s checkpoint cache, Job restarting constantly


Screenshot 2023-07-03 at 10 50 44

JM CPU/memory use

```
NAME                                         CPU(cores)   MEMORY(bytes)
flink-jobmanager-6797f4cdbd-xll86            70m          1237Mi
```


Screenshot 2023-07-03 at 10 47 53


### Test 3: No checkpoint cache, Job restarting constantly

Screenshot 2023-07-03 at 11 09 43

JM CPU/memory use

```
NAME                                         CPU(cores)   MEMORY(bytes)
flink-jobmanager-79f5768fdc-2t9hd            50m          1272Mi
```
 
Screenshot 2023-07-03 at 11 09 10

File: flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java [hunk 1]
```diff
@@ -193,6 +193,28 @@ public class RestOptions {
...
+     * specified using notation: "500 ms", "1 s".
+     */
+    @Documentation.Section(Documentation.Sections.EXPERT_REST)
+    public static final ConfigOption<Duration> CACHE_CHECKPOINT_STATISTICS_TIMEOUT =
+            key("rest.cache.checkpoint-statistics.timeout")
+                    .durationType()
+                    .defaultValue(Duration.ofSeconds(3))
+                    .withDescription(
+                            "Duration from write after which cached checkpoints statistics are cleaned up. It can"
+                                    + " be specified using notation: \"500 ms\", \"1 s\".");
```
Comment (dmvk) [inline flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java:207]: We shouldn't be explaining how to specify duration here. If something needs to be clarified, we should develop a generalized explainer because it applies to many fields.
```suggestion
                    .withDescription(
                            "Duration from write, after which cached checkpoints statistics are cleaned up.");
```

File: flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java [hunk 2]
```diff
@@ -193,6 +193,28 @@ public class RestOptions {
                                     + "Lowering the thread priority will give Flink's main components more CPU time whereas "
                                     + "increasing will allocate more time for the REST server's processing.");
 
+    /**
+     * Duration from write after which cached checkpoints statistics are cleaned up. It can be
+     * specified using notation: "500 ms", "1 s".
```
Comment (dmvk) [inline flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java:198]:
```suggestion
     * Duration from write, after which cached checkpoints statistics are cleaned up.
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 3]
```diff
@@ -937,6 +938,15 @@ public CompletableFuture<ExecutionGraphInfo> requestExecutionGraphInfo(
                 .exceptionally(checkExecutionGraphStoreOnException);
     }
 
+    @Override
+    public CompletableFuture<CheckpointStatsSnapshot> requestCheckpointStats(
+            JobID jobId, Time timeout) {
+        Optional<JobManagerRunner> maybeJob = getJobManagerRunner(jobId);
+
+        return maybeJob.map(job -> job.requestCheckpointStats(timeout))
+                .orElse(FutureUtils.completedExceptionally(new FlinkJobNotFoundException(jobId)));
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:947]: `org.apache.flink.runtime.dispatcher.Dispatcher#performOperationOnJobMasterGateway` should be used here instead

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java [hunk 4]
```diff
@@ -96,6 +97,15 @@ public interface JobManagerRunner extends AutoCloseableAsync {
...
     CompletableFuture<ExecutionGraphInfo> requestJob(Time timeout);
 
+    /**
+     * Requests the {@link CheckpointStatsSnapshot} of the executed job.
+     *
+     * @param timeout for the rpc call
+     * @return Future which is completed with the {@link CheckpointStatsSnapshot} of the executed
+     *     job. Returns null if job is not running.
+     */
+    CompletableFuture<CheckpointStatsSnapshot> requestCheckpointStats(Time timeout);
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java:107]: Why is this needed?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java [hunk 5]
```diff
@@ -21,17 +21,22 @@
...
 
 /** Configuration object containing values for the rest handler configuration. */
 public class RestHandlerConfiguration {
 
     private final long refreshInterval;
 
     private final int maxCheckpointStatisticCacheEntries;
+    private final Duration checkpointStatsSnapshotCacheExpireAfterWrite;
+
+    private final int checkpointStatsSnapshotCacheSize;
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java:39]: nit: the spacings here are rather inconsistent

**Comment by dmvk** [comment]
> Periodic 500 returned by /checkpoints on the first call after a job restart (likely that there are no "Quantile" statistics after only 1 checkpoint. This is consistent for all 3 tests, meaning its an independent issue.

Can you please open a JIRA issue describing this finding (if it doesn't already exist)? It sounds like something that should be fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java [hunk 6]
```diff
@@ -93,6 +94,8 @@ ExecutionState requestPartitionState(
 
     ExecutionGraphInfo requestJob();
 
+    CheckpointStatsSnapshot requestCheckpointStats();
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java:97]: Why do we need this? Can JM access this through the ExecutionGraphInfo?

**Review by dmvk — COMMENTED** [review]
I'm still bit struggling to grasp the motivation. Can you please try to expand on it [1]?

[1] https://issues.apache.org/jira/browse/FLINK-32469?focusedCommentId=17739832&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17739832

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/checkpoints/AbstractCheckpointStatsHandler.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.concurrent.Executor;
+
+/**
+ * Abstract class for checkpoint handlers that will cache the {@link CheckpointStatsSnapshot}
+ * object.
+ *
+ * @param <R> the response type
+ * @param <M> the message parameters
+ */
+public abstract class AbstractCheckpointStatsHandler<
```
Comment (dannycranmer) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/checkpoints/AbstractCheckpointStatsHandler.java:55]: Missing compatibility annotation

File: docs/layouts/shortcodes/generated/expert_rest_section.html [hunk 8]
```diff
@@ -20,6 +20,18 @@
...
+            <td style="word-wrap: break-word;">1000</td>
+            <td>Integer</td>
+            <td>Maximum number of entries in the checkpoint statistics cache.</td>
+        </tr>
+        <tr>
+            <td><h5>rest.cache.checkpoint-statistics.timeout</h5></td>
+            <td style="word-wrap: break-word;">3 s</td>
+            <td>Duration</td>
+            <td>Duration from write after which cached checkpoints statistics are cleaned up. It can be specified using notation: "500 ms", "1 s".</td>
+        </tr>
```
Comment (dannycranmer) [inline docs/layouts/shortcodes/generated/expert_rest_section.html:34]: Are these configs backwards compatible? Seems like `rest.cache.checkpoint-statistics.timeout` is a new config split for `web.refresh-interval`?

File: docs/layouts/shortcodes/generated/expert_rest_section.html [hunk 9]
```diff
@@ -20,6 +20,18 @@
...
+            <td><h5>rest.cache.checkpoint-statistics.size</h5></td>
+            <td style="word-wrap: break-word;">1000</td>
+            <td>Integer</td>
+            <td>Maximum number of entries in the checkpoint statistics cache.</td>
+        </tr>
+        <tr>
+            <td><h5>rest.cache.checkpoint-statistics.timeout</h5></td>
+            <td style="word-wrap: break-word;">3 s</td>
+            <td>Duration</td>
+            <td>Duration from write after which cached checkpoints statistics are cleaned up. It can be specified using notation: "500 ms", "1 s".</td>
```
Comment (dannycranmer) [inline docs/layouts/shortcodes/generated/expert_rest_section.html:33]:
```suggestion
            <td>Duration from write after which cached checkpoints statistics are cleaned up.</td>
```

File: docs/layouts/shortcodes/generated/rest_configuration.html [hunk 10]
```diff
@@ -38,6 +38,18 @@
...
+            <td><h5>rest.cache.checkpoint-statistics.size</h5></td>
+            <td style="word-wrap: break-word;">1000</td>
+            <td>Integer</td>
+            <td>Maximum number of entries in the checkpoint statistics cache.</td>
+        </tr>
+        <tr>
+            <td><h5>rest.cache.checkpoint-statistics.timeout</h5></td>
+            <td style="word-wrap: break-word;">3 s</td>
+            <td>Duration</td>
+            <td>Duration from write after which cached checkpoints statistics are cleaned up. It can be specified using notation: "500 ms", "1 s".</td>
```
Comment (dannycranmer) [inline docs/layouts/shortcodes/generated/rest_configuration.html:51]:
```suggestion
            <td>Duration from write after which cached checkpoints statistics are cleaned up.</td>
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java [hunk 11]
```diff
@@ -46,6 +51,8 @@ public class RestHandlerConfiguration {
     public RestHandlerConfiguration(
             long refreshInterval,
             int maxCheckpointStatisticCacheEntries,
+            Duration checkpointStatsSnapshotCacheExpireAfterWrite,
+            int checkpointStatsSnapshotCacheSize,
```
Comment (dannycranmer) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java:55]: It is confusing to understand the difference between `maxCheckpointStatisticCacheEntries` and `checkpointStatsSnapshotCacheSize` without reading the code. Can the variable names be improved?

**Review by dannycranmer — CHANGES_REQUESTED** [review]

**Comment by hlteoh37** [comment]
> > Periodic 500 returned by /checkpoints on the first call after a job restart (likely that there are no "Quantile" statistics after only 1 checkpoint. This is consistent for all 3 tests, meaning its an independent issue.
> 
> Can you please open a JIRA issue describing this finding (if it doesn't already exist)? It sounds like something that should be fixed.

Thats a good callout. Created https://issues.apache.org/jira/browse/FLINK-32535

File: flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java (hunk 1 above)
Comment (hlteoh37) [reply flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java:207]: Good point. Will incorporate feedback

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java (hunk 4 above)
Comment (hlteoh37) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java:107]: This particular change is required if we want to just request the `CheckpointStatsSnapshot` object from the JobMaster to side step the recursive copy of the ArchivedExecutionGraph on the JM side.

Currently, the process is:
Client -> REST (cache ExecutionGraph) -> JobManager -> JobMaster (builds ArchivedExecutionGraph - recursive copy, costly(?))

Without this change, the process will be
Client -> REST (cache CheckpointStatsSnapshot) -> JobManager (request ExecutionGraphInfo) -> JobMaster (builds ArchivedExecutionGraph - recursive copy, costly (?))

With this change, the process will be
Client -> REST (cache CheckpointStatsSnapshot) -> JobManager (request CheckpointStatsSnapshot) -> JobMaster (retrieves CheckpointStatsSnapshot from cache in CheckpointStatsTracker)


I think this change can be helpful to refactor the `JobManager` / `JobMaster` API to retrieve specific information about a job, rather than 1 large data dump (ExecutionGraph)! Let me know what you think

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java (hunk 5 above)
Comment (hlteoh37) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java:39]: Ok fixed

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java (hunk 11 above)
Comment (hlteoh37) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java:55]: Ok renamed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/checkpoints/AbstractCheckpointStatsHandler.java (hunk 7 above)
Comment (hlteoh37) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/checkpoints/AbstractCheckpointStatsHandler.java:55]: Good callout. 

This seems to be missing for all REST API classes. I've used @Internal for this class, and created a followup JIRA https://issues.apache.org/jira/browse/FLINK-32537

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java (hunk 6 above)
Comment (hlteoh37) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java:97]: Same as above, Yes JM can access this via ExecutionGraphInfo, but that means we would incur the cost of recursive copy of the ExecutionGraph when all we need is the CheckpointStatsSnapshot object.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java (hunk 4 above)
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java:107]: The point is that you don't need to introduce a new method in JobManagerRunner, but instead, Dispatcher can directly talk to the JobMasterGateway (`performOperationOnJobMasterGateway` is what you should look for).

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java (hunk 6 above)
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java:97]: 👍 can you please add a javadoc with an explanation?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java (hunk 4 above)
Comment (hlteoh37) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunner.java:107]: Ah. Got it! Will refactor

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java (hunk 6 above)
Comment (hlteoh37) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java:97]: Yep can do.

File: docs/layouts/shortcodes/generated/expert_rest_section.html (hunk 8 above)
Comment (hlteoh37) [reply docs/layouts/shortcodes/generated/expert_rest_section.html]: Ok, made it such that it will fallback to `web.refresh-interval`

**Comment by hlteoh37** [comment]
Have addressed all comments, would appreciate a re-review if you have time, @dmvk @dannycranmer !

**Comment by hlteoh37** [comment]
Added doc strings and fixed a typo

**Comment by hlteoh37** [comment]
@dmvk @dannycranmer  I've addressed all the comments so far - any further comments on this PR? Looking to move this forward!

File: docs/layouts/shortcodes/generated/expert_rest_section.html [hunk 12]
```diff
@@ -20,6 +20,18 @@
             <td>Long</td>
             <td>The time in ms that the client waits for the leader address, e.g., Dispatcher or WebMonitorEndpoint</td>
         </tr>
+        <tr>
+            <td><h5>rest.cache.checkpoint-statistics.size</h5></td>
```
Comment (dannycranmer) [inline docs/layouts/shortcodes/generated/expert_rest_section.html:24]: Cn you justify why this config should go into the `rest.*` bucket and not the `web.*` bucket? Seems llike we have similar configs in the `web.*` section, for example `web.exception-history-size`

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: docs/layouts/shortcodes/generated/expert_rest_section.html (hunk 12 above)
Comment (hlteoh37) [reply docs/layouts/shortcodes/generated/expert_rest_section.html:24]: In my mind, `rest.*` should be configs that are REST API relation, whereas `web.*` should control configs that are Flink dashboard related (e.g. `web.refresh-interval`. 

Decided to go with `rest.*` because this cache is a REST API concept

File: flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java [hunk 13]
```diff
@@ -193,6 +193,24 @@ public class RestOptions {
                                     + "Lowering the thread priority will give Flink's main components more CPU time whereas "
                                     + "increasing will allocate more time for the REST server's processing.");
 
+    /** Duration from write, after which cached checkpoints statistics are cleaned up. */
+    @Documentation.Section(Documentation.Sections.EXPERT_REST)
+    public static final ConfigOption<Duration> CACHE_CHECKPOINT_STATISTICS_TIMEOUT =
+            key("rest.cache.checkpoint-statistics.timeout")
+                    .durationType()
+                    .noDefaultValue()
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java:202]: Good point. Will override the default description with `@OverrideDefault`

File: docs/layouts/shortcodes/generated/expert_rest_section.html [hunk 14]
```diff
@@ -20,6 +20,18 @@
...
+            <td style="word-wrap: break-word;">1000</td>
+            <td>Integer</td>
+            <td>Maximum number of entries in the checkpoint statistics cache.</td>
+        </tr>
+        <tr>
+            <td><h5>rest.cache.checkpoint-statistics.timeout</h5></td>
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>Duration</td>
+            <td>Duration from write after which cached checkpoints statistics are cleaned up. For backwards compatibility, if this is not specified, the web.refresh-interval will be used instead.</td>
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply docs/layouts/shortcodes/generated/expert_rest_section.html:33]: Ok, updated

**Comment by hlteoh37** [comment]
@flinkbot run azure

File: flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-core/src/main/java/org/apache/flink/configuration/RestOptions.java:202]: Created a followup JIRA
https://issues.apache.org/jira/browse/FLINK-32619

**Comment by dannycranmer** [comment]
I will merge tomorrow unless there is additional feedback
